protocol/deployment-chains.md

```diff
@@ -0,0 +1,261 @@
+So it is not a single deployment sequence, but more like a tree of possible deployments.
+
+##### How do we keep the cache reasonably sized?
+
+Produced increments might take up more space than desired:
+for simplicity, dumping the whole state-allocs each step would be nice.
+
+But if there are many configuration tweaks, and many steps, it may produce too many of these files.
+
+The devnet L1 allocs file today is ~650 KB.
+The devnet L2 allocs file today is 9.1 MB: the majority of this data are L2-predeploy proxy placeholders.
+
+To reduce the size we could make the output incremental also:
+an output can be JSON, and each account entry overwrites the contents of the previous account entry.
+For simplicity we avoid storage-merging, and just overwrite accounts as a whole.
+We do not have contracts with particularly many storage entries, nor do the medium-sized ones change with every step. 
+
+By performing merges this way, we can also make hardforks clean:
+e.g. we can produce a state for fork A, and then fork it and apply the upgrade to fork B, producing a tiny diff.
+
+The smaller the files, the more we can afford, and the higher the limit
+for configuration test-coverage is.
+
+##### How do we not miss the cache with every redeployment?
+
+It is important to omit the "genesis timestamp" from the cache-key, and use some fixed value as input instead:
+every real-time test will need a different timestamp, so embedding it as input would break all caching.
+Luckily, only the legacy L2-output-oracle L1 contract enshrines anything,
+and the L2 genesis block is easily modified before chain-specs are generated.
+
+##### How does this work with production deployments?
+
+The caching, chaining, and commiting is not a part of production.
+As it is against a live chain, not a local state-dump.
+
+However, the isolated configuration and deployment-sub-routines do enable
+a deployer to plug into the superchain contracts, while customizing just what they need,
+without having to untangle a monolithic single deployment script.
+
+Ideally this is performed through the OPStackManager.
+
+### Result
+
+This approach enables us to get rid of "mega-feature" env-vars (Plasma, fault-proofs, Interop)
+as well as allocs hardfork-suffixes: everything is configuration,
+and we can just fork of existing configuration to tweak something.
+
+The main changes to the deploy/genesis-scripting are to:
+1. Get rid of L1 and L2 "allocs": every incremental output is a small "allocs", identified by the hash of inputs.
+2. Get rid of env-vars: we make all customization the same type of well-defined input.
+3. Split the deploy-config: each deployment sub-routine should have its own well-contained config,
+  such that we can attribute the output to the change of input.
+
+The Go `op-e2e` side should then be able to produce the needed configuration input-groups, hash them,
+and then either hit the cache of outputs, or invoke the solidity-scripts to generate the outputs. 
```

Comment:
One thing to be careful about is how we handle the contract bytecode in genesis, and how to release it.

Currently, because of our monorepo structure, node releases (which encompass predeploys via genesis) and contracts releases (which to date have been only scoped to L1 contracts) are independently versioned and released. We want to make sure that as we change/add predeploys:
1. They are clearly released as either part of the node releases, the contract releases, or a new release set. We should decide on their release process and document it in the monorepo READMEs or [VERSIONING.md](https://github.com/ethereum-optimism/optimism/blob/develop/packages/contracts-bedrock/VERSIONING.md)
2. When generating genesis files for development, we can always use the latest predeploy contract bytecode. But when generating genesis files for production, we must ensure only released versions are used.

Reply:
This is all cached by content-hash, and testing is always on latest develop, the versioning part is out of scope.